### PR TITLE
Ana/fix chunk errors for real

### LINKED ISSUE #3760: Let FE know if it is outdated
**Is your feature request related to a problem? Please describe.**

Sometimes our FE is outdated as the files keep being cached.

**Describe the solution you'd like**

Based on:
https://medium.com/hypefactors/busting-spa-chunk-cache-for-active-users-after-a-deploy-a9a06242edb1

We may do this as we already talked about once:
- [ ] on deploy in Netlify trigger a mutation in Apollo to write in the table `versions` a record for the latest deployed version in `production` and ` staging`
- [ ] clients load/subscribe to the version
- [ ] clients show a reload message if there is new version while the app is open
- [ ] clicking on the outdated warning should force refresh the app
- [ ] clients refresh when the version is different then the current version on inital load

**Describe alternatives you've considered**

We already tried shipping the version together with the app on `/version.json` which has similar caching issues.

We can also figure out, why the index.html is not properly rechecked on every load.

**Additional context**

### PR DESCRIPTION
**Pull request by Bitcoinera**
Closes #3760

**Description:**

Still work in progress although not looking bad

Thank you! 🚀

---

For contributor:

- [ ] Added changes entries. Run `yarn changelog` for a guided process.
- [ ] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

File: tasks/poll-index.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    .then(result => {
+        if (!loaded) {
+            currentEtag = result.headers._headers.etag[0]
+            loaded = true
+        }
+        if (currentEtag !== result.headers._headers.etag[0]) {
+            console.log('Outdated Lunie version detected! Refreshing...')
+            // the client has an outdated version. We need to refresh the browser to get the latest changes (and chunks)
+            // when the reload method receives true as an argument, it will always reload the page from the server
+            window.location.reload(true)
```
Comment (Bitcoinera) [inline tasks/poll-index.js:29]: Here is where I have some doubts... Is the `window` object reachable here?

Comment (faboweb) [reply tasks/poll-index.js:29]: on web always

File: tasks/poll-index.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+/*
+ * this script polls the Lunie site every 10 seconds and extracts its etag to check if there has been 
+ * any new Lunie versions deployed on Netlify while the user had the app open on the browser
+ * 
+ * Usage:
+ *
+ * node tasks/poll-index.js
+ */
+
+const fetch = require("node-fetch")
```
Comment (faboweb) [inline tasks/poll-index.js:10]: why node fetch?

File: tasks/poll-index.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+/*
```
Comment (faboweb) [inline tasks/poll-index.js:1]: this should not go into tasks as it is part of the app, right?

File: tasks/poll-index.js (hunk 2 above)
Comment (Bitcoinera) [reply tasks/poll-index.js:10]: It is the one we are using in the other scripts 🤷‍♀

File: tasks/poll-index.js (hunk 3 above)
Comment (Bitcoinera) [reply tasks/poll-index.js:1]: But it should be a separate node process running along in the same instance right?

File: tasks/poll-index.js (hunk 2 above)
Comment (faboweb) [reply tasks/poll-index.js:10]: lol yeah but they run in Node.js. This runs in the browser where `fetch` is native

File: tasks/poll-index.js (hunk 3 above)
Comment (faboweb) [reply tasks/poll-index.js:1]: noooo, this runs in the browser not in node

Comment (faboweb) [reply tasks/poll-index.js:1]: or do you want to run it in the API?

Comment (faboweb) [reply tasks/poll-index.js:1]: we could also run this in the API and on an update trigger an apollo subscription

Comment (Bitcoinera) [reply tasks/poll-index.js:1]: no, I think in the browser it makes more sense & it is more simple

File: tasks/poll-index.js (hunk 2 above)
Comment (Bitcoinera) [reply tasks/poll-index.js:10]: oh I see 😮

Comment (Bitcoinera) [reply tasks/poll-index.js]: I tried removing it and it didn't work 🤷‍♀ it said `fetch undefined`

Comment (faboweb) [reply tasks/poll-index.js]: Ur still executing the script with node...

**Comment by Bitcoinera** [comment]
I think at this point it should work... problem is that https://app.lunie.io does not have the Access-Control-Allow-Origin header, so until then I cannot make requests there.

That would be how it would work. I can try and change the request to the deploy itself just to test

**Comment by Bitcoinera** [comment]
I certainly don't know how to test this...

Using the preview url it gets automatically translated to https://app.lunie.io. Without the headers set in production, it seems impossible to test this 🤷‍♀

File: public/_headers [hunk 4]
```diff
@@ -2,6 +2,7 @@
   X-Frame-Options: DENY
   X-XSS-Protection: 1; mode=block
   Content-Security-Policy: frame-ancestors 'self'
+  Access-Control-Allow-Origin: *
```
Comment (faboweb) [inline public/_headers:5]: is that needed?

File: src/scripts/check-for-new-lunie-versions.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
+/*
+ * this script polls the Lunie site every minute and extracts its etag to check if there has been
+ * any new Lunie versions deployed on Netlify while the user had the app open on the browser
+ *
+ * Usage:
+ *
+ * node tasks/poll-index.js
+ */
+
+const fetch = require("node-fetch")
```
Comment (faboweb) [inline src/scripts/check-for-new-lunie-versions.js:10]:
```suggestion
```

File: src/scripts/check-for-new-lunie-versions.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
+/*
+ * this script polls the Lunie site every minute and extracts its etag to check if there has been
+ * any new Lunie versions deployed on Netlify while the user had the app open on the browser
+ *
+ * Usage:
+ *
+ * node tasks/poll-index.js
```
Comment (faboweb) [inline src/scripts/check-for-new-lunie-versions.js:7]:
```suggestion
```

File: src/scripts/check-for-new-lunie-versions.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        currentEtag = result.headers._headers.etag[0]
+        loaded = true
+      }
+      console.log("Client Etag", currentEtag)
+      if (currentEtag !== result.headers._headers.etag[0]) {
+        console.log("Outdated Lunie version detected! Refreshing...")
+        // the client has an outdated version. We need to refresh the browser to get the latest changes (and chunks)
+        // when the reload method receives true as an argument, it will always reload the page from the server
+        window.location.reload(true)
+        loaded = false
```
Comment (faboweb) [inline src/scripts/check-for-new-lunie-versions.js:29]:
```suggestion
```

File: public/_headers (hunk 4 above)
Comment (Bitcoinera) [reply public/_headers:5]: it seems so

File: src/scripts/check-for-new-lunie-versions.js (hunk 7 above)
Comment (Bitcoinera) [reply src/scripts/check-for-new-lunie-versions.js:29]: Then it won't update the client etag

File: src/scripts/check-for-new-lunie-versions.js (hunk 5 above)
Comment (Bitcoinera) [reply src/scripts/check-for-new-lunie-versions.js:10]: Yeah, you are right. This was because I was testing at the beginning with node 🤦‍♀

**Comment by Bitcoinera** [comment]
I cannot go past this. If I try using the preview url, it directly translates it to https://app.lunie.io...

image

Any idea how to work around this?

My solution would be to merge the _headers with the Access-Control-Allow-Origin header into production and then test this

File: public/_headers (hunk 4 above)
Comment (faboweb) [reply public/_headers:5]: noooooo

File: src/scripts/check-for-new-lunie-versions.js (hunk 7 above)
Comment (faboweb) [reply src/scripts/check-for-new-lunie-versions.js:23]: why? `window.location.reload(true)` will reload the whole page

**Comment by faboweb** [comment]
don't specify an URL at all. just get `/index.html` it will resolve to the current URL + `/index.html`

**Comment by faboweb** [comment]
If this doesn't work, you can use `window.location.hostname`

**Comment by faboweb** [comment]
querying app.lunie.io is also bad for previews as you would compare your local build against the production build

**Comment by Bitcoinera** [comment]
Still getting the very same CORS error. And I already tried using the "/index.html" solution before.

The issue is that we don't have that header set in production and that the preview URL is directly transalated to https://app.lunie.io for some reason

**Comment by faboweb** [comment]
Did you test using window.location.host?

**Comment by Bitcoinera** [comment]
> Did you test using window.location.host?

Wow, this was it!

**Comment by Bitcoinera** [comment]
It seems it didn't work so well window.location.host... Look at the URL it is calling 👀 

image

**Comment by Bitcoinera** [comment]
I think I am getting an opaque response due to CORS... `headers` is just shown as a prototype 🤔

**Comment by Bitcoinera** [comment]
Na, will look more into this tomorrow. Getting that opaque response

**Comment by Bitcoinera** [comment]
Amazing, this works! Good bye chunk errors!

I'm so happy and have learnt a lot along the way! 🤩

File: src/initializeApp.js [hunk 8]
```diff
@@ -34,6 +35,8 @@ function setOptions(urlParams, store) {
 export default async function init(urlParams, env = process.env) {
   // add error handlers in production
   if (env.NODE_ENV === `production`) {
+    // check every minute if new Lunie versions have been deployed
```
Comment (jbibla) [inline src/initializeApp.js:38]: what do you mean "every minute"

File: public/_headers (hunk 4 above)
Comment (faboweb) [reply public/_headers:5]: so we can remove this, right?

File: src/scripts/check-for-new-lunie-versions.js (hunk 7 above)
Comment (Bitcoinera) [reply src/scripts/check-for-new-lunie-versions.js:26]: oh, and then it reruns the whole function making this unnecessary you mean? 🤔 

Makes sense...

File: public/_headers (hunk 4 above)
Comment (Bitcoinera) [reply public/_headers:5]: Maybe you can give a better explanation than "nooooo" 😅 

I will try

**Comment by Bitcoinera** [comment]
OK, @faboweb applied your suggestions.

Let's merge this!

**Comment by Bitcoinera** [comment]
Why is pendingUpdated failing so much on this PR?

**Comment by faboweb** [comment]
> Why is pendingUpdated failing so much on this PR?

because there is no changelog entry

**Comment by Bitcoinera** [comment]
xD I would have sworn I had added it! 37 commits and no changelog ..

File: tests/e2e/helpers.js [hunk 9]
```diff
@@ -252,8 +252,8 @@ async function getAccountBalance(browser) {
         browser.globals.denom = total[1]
         browser.globals.totalAtoms = total[0]
       })
-      /*await browser.getText(".total-atoms h2", result => {
-        browser.globals.totalAtoms = result.value.replace(",", "")
+      /*await browser.getText(".total-atoms h2", result => {
+        browser.globals.totalAtoms = result.value.replace(",", "")
       })*/
```
Comment (faboweb) [inline tests/e2e/helpers.js:257]:
```suggestion
```

**Pull request merged**